semmc/tests/semstore/TestArch.hs

```diff
@@ -0,0 +1,196 @@
+  testing to be performed.  Not a real architecture at all... not even
+  a little bit.
+|-}
+
+module TestArch
+where
+
+import           Data.EnumF -- in Dismantle Tablegen!
+import           Data.Kind ( Type )
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.HasRepr as HR
+import qualified Data.Parameterized.List as PL
+import           Data.Parameterized.Some
+import qualified Data.Parameterized.SymbolRepr as SR
+import qualified Data.Parameterized.TH.GADT as TH
+import qualified Data.Text as T
+import           GHC.TypeLits ( Symbol )
+import           Numeric.Natural
+import qualified SemMC.Architecture as SA
+import qualified SemMC.Architecture.Location as L
+import           What4.BaseTypes
+
+
+data TestGenArch  -- ^ the architecture type for testing
+
+instance SA.Architecture TestGenArch where
+  data TaggedExpr TestGenArch sym s =
+    TaggedExpr { unTaggedExpr :: SA.AllocatedOperand TestGenArch sym s }
+
+  unTagged te = case unTaggedExpr te of
+    SA.LocationOperand _ se -> Just se
+
+  taggedOperand = unTaggedExpr
+
+  operandToLocation _ FooArg     = Nothing
+  operandToLocation _ BarArg     = Just TestBarLoc
+  operandToLocation _ (BoxArg n) = Just (TestBoxLoc n)
+
+  uninterpretedFunctions _ = []  -- TODO: add some
+
+  shapeReprToTypeRepr _ sr =
+    case SR.symbolRepr sr of
+      "Foo"
+        | Just Refl <- testEquality sr (SR.knownSymbol @"Foo")
+          -> BaseNatRepr
+      "Bar"
+        | Just Refl <- testEquality sr (SR.knownSymbol @"Bar")
+          -> BaseBVRepr (knownNat :: NatRepr 32)
+      "Box"
+        | Just Refl <- testEquality sr (SR.knownSymbol @"Box")
+          -> BaseBVRepr (knownNat :: NatRepr 32)
+      _ -> error ("Invalid shape repr: " ++ show sr)
+
+  allocateSymExprsForOperand _arch _sym newVars FooArg = undefined
+  allocateSymExprsForOperand _arch _sym newVars BarArg =
+    let loc = TestBarLoc in
+    TaggedExpr <$> SA.LocationOperand loc <$> newVars loc
+  allocateSymExprsForOperand _arch _sym newVars (BoxArg n) =
+    let loc = TestBoxLoc n in
+    TaggedExpr <$> SA.LocationOperand loc <$> newVars loc
+
+  locationFuncInterpretation _ = []
+
+
+----------------------------------------------------------------------
+-- Location
+
+data TestLocation :: BaseType -> Type where
+  TestNatLoc :: Natural -> TestLocation BaseNatType
+  TestIntLoc :: Integer -> TestLocation BaseIntegerType
+  TestBoxLoc :: Natural -> TestLocation (BaseBVType 32)
+  TestBarLoc :: TestLocation (BaseBVType 32)
+  -- TBD: more basetype locations
+  -- TBD: some memory locations
+  -- MemLoc :: Mem -> Location (BaseBVType 32)
+
+instance Show (TestLocation tp) where
+  show TestBarLoc     = "Bar"
+  show (TestBoxLoc n) = "Box_" <> show n
+  show (TestNatLoc n) = "NAT_" <> show n  -- KWQ: want NAT@... see above
+  show (TestIntLoc i) = if i >= 0
+                        then "INT_" <> show i  -- KWQ: want INT@... see above
+                        else "NEGINT_" <> show (-i)
```

Comment:
I wasn't able to figure out what the "see above" comment was referring to - are those strings already used for something else?

Reply:
Whoops, I removed the original comment, which was on the TestBoxLoc show.  Basically it was a note that I originally had something like "Box#0", but that isn't a valid identifier (and for these it was going to be "NAT@{natval}").  I was initially questioning whether we needed to consider quoting these or alternatively detecting and rejecting them, but they should just be internal identifiers and we can control the use, so I think just removing the comments is the best route here.